### PR TITLE
Ignore Volume if similar mountpath found

### LINKED ISSUE #544: Fatal error saying "/etc/nginx/certs": must be unique
I'm using `jrcs/letsencrypt-nginx-proxy-companion` in my `docker-compose.production.yml` and `kompose -f docker-compose.production.yml up` throws an error saying

```
WARN Unsupported logging key - ignoring
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
INFO We are going to create Kubernetes Deployments, Services and PersistentVolumeClaims for your Dockerized application. If you need different kind of resources, use the 'kompose convert' and 'kubectl create -f' commands instead.

INFO Successfully created Service: letsencrypt
INFO Successfully created Service: proxy
FATA Error while deploying application: Deployment.extensions "letsencrypt" is invalid: spec.template.spec.containers[0].volumeMounts[4].mountPath: Invalid value: "/etc/nginx/certs": must be unique
```

How can I fix this issue?

---

My `kompose version` is `0.5.0 ()`

And my `docker-compose.production.yml` below.

```yml
version: "2"

services:

  proxy:
    image: jwilder/nginx-proxy
    privileged: true
    container_name: proxy
    ports:
      - 80:80
      - 443:443
    volumes:
      - /etc/nginx/vhost.d
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./certs:/etc/nginx/certs:ro
      - /usr/share/nginx/html
    restart: always
    logging:
      options:
        max-size: 5m
        max-file: "10"

  letsencrypt:
    image: jrcs/letsencrypt-nginx-proxy-companion
    privileged: true
    container_name: ssl
    volumes_from:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - ./certs:/etc/nginx/certs:rw

networks:
  default:
    external:
      name: shared
```

### PR DESCRIPTION
**Pull request by surajnarwade**
Resolves #544

With this PR, kompose will give error if similar mountpaths are found
in case when `volumes_from` is being used.

### CODE REVIEW

**Comment by surajnarwade** [comment]
cc @procrypt @cdrage @surajssd @containscafeine

File: pkg/transformer/kubernetes/k8sutils.go [hunk 1]
```diff
@@ -497,6 +497,10 @@ func (k *Kubernetes) VolumesFrom(objects *[]runtime.Object, komposeObject kobjec
 				if err != nil {
 					return errors.Wrap(err, "k.findDependentVolumes")
 				}
+				err = checkSimilarMount(volumeMounts, t.Spec.Template.Spec.Containers[0].VolumeMounts)
+				if err != nil {
+					log.Fatal(err)
```
Comment (surajssd) [inline pkg/transformer/kubernetes/k8sutils.go]: before every fatal out can we should mention which service has this problem? So it's easy where ther is a repetitive volume Mount?

**Comment by surajnarwade** [comment]
@surajssd done

**Comment by cdrage** [comment]
@surajnarwade Need tests before merge

**Comment by surajnarwade** [comment]
@cdrage added test.

File: pkg/transformer/kubernetes/k8sutils.go [hunk 2]
```diff
@@ -527,10 +539,26 @@ func (k *Kubernetes) VolumesFrom(objects *[]runtime.Object, komposeObject kobjec
...
+				}
 				t.Spec.Template.Spec.Volumes = append(t.Spec.Template.Spec.Volumes, volumes...)
 				t.Spec.Template.Spec.Containers[0].VolumeMounts = append(t.Spec.Template.Spec.Containers[0].VolumeMounts, volumeMounts...)
 			}
 		}
 	}
 	return nil
 }
+
+func CheckSimilarMount(depVolumeMounts []api.VolumeMount, volumeMounts []api.VolumeMount, svcName string) error {
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: Please add comments on what this does, it's functionality, etc.

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 3]
```diff
@@ -275,3 +275,32 @@ func TestRecreateStrategyWithVolumesPresent(t *testing.T) {
 		}
 	}
 }
+
+func TestCheckSimilarMount(t *testing.T) {
+	//var qw []api.VolumeMount{}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils_test.go]: Forgotten comment?

**Comment by surajnarwade** [comment]
@cdrage added comments

File: pkg/transformer/kubernetes/k8sutils.go [hunk 4]
```diff
@@ -527,10 +539,27 @@ func (k *Kubernetes) VolumesFrom(objects *[]runtime.Object, komposeObject kobjec
...
 	}
 	return nil
 }
+
+//It will check for similar mounts if 'volumes_from' key is used, if found, it will give error
+func CheckSimilarMount(depVolumeMounts []api.VolumeMount, volumeMounts []api.VolumeMount, svcName string) error {
+	for _, depVolumeMount := range depVolumeMounts {
+		for _, volumeMount := range volumeMounts {
+			if depVolumeMount.MountPath == volumeMount.MountPath {
+				return errors.Errorf("In %q, similar mountpath found %q: must be unique", svcName, volumeMount.MountPath)
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: A better output may be:
```
In service %q, a similar mount-path was found '%s'. Mount-paths in Kubernetes must be unique.
```

File: pkg/transformer/kubernetes/k8sutils.go [hunk 5]
```diff
@@ -527,10 +539,27 @@ func (k *Kubernetes) VolumesFrom(objects *[]runtime.Object, komposeObject kobjec
 				if err != nil {
 					return errors.Wrap(err, "k.findDependentVolumes")
 				}
+				err = CheckSimilarMount(volumeMounts, t.Spec.Template.Spec.Containers[0].VolumeMounts, t.Name)
+				if err != nil {
+					log.Fatal(err)
+				}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils.go]: I wish we could refactor all these functions, but that'd be for another day..

File: pkg/transformer/kubernetes/k8sutils_test.go [hunk 6]
```diff
@@ -275,3 +275,33 @@ func TestRecreateStrategyWithVolumesPresent(t *testing.T) {
 		}
 	}
 }
+
+//Test for checking similar mountpath if 'volumes_from' key is used
+func TestCheckSimilarMount(t *testing.T) {
+	//var qw []api.VolumeMount{}
```
Comment (cdrage) [inline pkg/transformer/kubernetes/k8sutils_test.go]: Comment is still here... Need to remove this.

**Comment by kadel** [comment]
I don't think that this is right solution here.

Following docker-compose file is completely valid with docker-compose :

```yaml
version: '2'

services:
  foo:
    image: busybox
    command: sleep 3600
    volumes:
       - /asdf

  bar:
    image: busybox
    command: sleep 3600
    volumes:
       - /asdf
    volumes_from:
       - foo
  ```

result is that `/asdf` volume is shared between both containers.

We should do same thing with kubernetes. 
We already using shared volumes for `volume_from`.

if container `bar` is mounting volumes from `foo` and one of the volumes has same path as is defined in bar `volumes` array we should ignore it , that will create same behavior as with docker-compose.

**Comment by surajnarwade** [comment]
cc @kadel @cdrage

**Comment by surajssd** [comment]
I am trying to convert the example @kadel has given. Since `/asdf` should be same volumes, so only one `pvc` should be created.

But after I convert I see two `pvc` created:

```bash
$ kompose convert 
[SNIP]
INFO file "bar-claim0-persistentvolumeclaim.yaml" created 
INFO file "foo-deployment.yaml" created           
INFO file "foo-claim0-persistentvolumeclaim.yaml" created 
```

Also see the `Deployment` it's not sharing the volumes.

```yaml
$ cat bar-deployment.yaml 
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  name: bar
spec:
  replicas: 1
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        io.kompose.service: bar
    spec:
      containers:
      - image: busybox
        name: bar
        resources: {}
        volumeMounts:
        - mountPath: /asdf
          name: bar-claim0
      restartPolicy: Always
      volumes:
      - name: bar-claim0
        persistentVolumeClaim:
          claimName: bar-claim0
status: {}
```
See the mount from above command it is coming from `pvc` `bar-claim0`

and in the below example the mount is coming from `pvc` `foo-claim0`.


```yaml
$ cat foo-deployment.yaml 
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  name: foo
spec:
  replicas: 1
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        io.kompose.service: foo
    spec:
      containers:
      - image: busybox
        name: foo
        resources: {}
        volumeMounts:
        - mountPath: /asdf
          name: foo-claim0
      restartPolicy: Always
      volumes:
      - name: foo-claim0
        persistentVolumeClaim:
          claimName: foo-claim0
status: {}
```

Rather they should be using same `pvc`.

**Comment by kadel** [comment]
I'm afraid that we can't do it as it's currently implemented :-( 

This will require bigger refactoring of how we handle volumes.

**Comment by surajnarwade** [comment]
I am closing this PR in favour of this PR, https://github.com/kubernetes/kompose/pull/626

**Pull request closed without merging**